library/core/src/cell.rs

```diff
@@ -1766,15 +1766,18 @@ impl<T: ?Sized + fmt::Display> fmt::Display for RefMut<'_, T> {
 ///
 /// The precise Rust aliasing rules are somewhat in flux, but the main points are not contentious:
 ///
-/// - If you create a safe reference with lifetime `'a` (either a `&T` or `&mut T`
-/// reference) that is accessible by safe code (for example, because you returned it),
-/// then you must not access the data in any way that contradicts that reference for the
-/// remainder of `'a`. For example, this means that if you take the `*mut T` from an
-/// `UnsafeCell<T>` and cast it to an `&T`, then the data in `T` must remain immutable
-/// (modulo any `UnsafeCell` data found within `T`, of course) until that reference's
-/// lifetime expires. Similarly, if you create a `&mut T` reference that is released to
-/// safe code, then you must not access the data within the `UnsafeCell` until that
-/// reference expires.
+/// - If you create a safe reference with lifetime `'a` (either a `&T` or `&mut T` reference), then
+/// you must not access the data in any way that contradicts that reference for the remainder of
+/// `'a`. For example, this means that if you take the `*mut T` from an `UnsafeCell<T>` and cast it
+/// to an `&T`, then the data in `T` must remain immutable (modulo any `UnsafeCell` data found
+/// within `T`, of course) until that reference's lifetime expires. Similarly, if you create a `&mut
+/// T` reference that is released to safe code, then you must not access the data within the
+/// `UnsafeCell` until that reference expires.
+///
+/// - For both `&T` without `UnsafeCell<T>` and `&mut T`, you must also not deallocate the data
+/// until the reference expires. As a special exception, `&UnsafeCell<T>` references may be
+/// deallocated during their lifetime. (However, whenever a `&UnsafeCell<T>` is constructed or
+/// dereferenced, it must still point to live memory.)
```

Comment:
This does become rather fragile once you consider larger structs that contain `UnsafeCell` fields: even if all fields of the struct use `UnsafeCell`, there might still be padding that would not be "inside" any `UnsafeCell` and hence not be allowed to be deallocated. This happened in https://github.com/rust-lang/rust/pull/98504#issuecomment-1166545014.

Reply:
So, in order to somehow tackle these situations without too much _ad-hoc_ `unsafe` that would thus need review, I guess the approach would be to define something along the lines of the following helper:

```rs
#[repr(transparent)]
struct ProtectorLessRef<'lt, Referee : ?Sized>(
    &'lt ::core::cell::SyncUnsafeCell<Referee>,
);
```

with:

  - some approapriate `Deref` impls (to avoid requiring `unsafe` on every deref or whatnot);
  - then maybe even a `core::ops::Receiver` impl (to allow usage with `self` methods);
  - and finally even maybe a touch of a type alias, _à la_`:
    ```rust
    // pseudo-code, but with helper traits this can be expressed:
    type ProtectorLess<&'lt T> = ProtectorLessRef<'lt, T>;
    ```

would allow writing stuff such as:

```rust
fn method(self: ProtectorLess<&'_ Self>, …) {
    …
    self.stuff();
    some_shared_ref_count.fetch_sub(1)…
    /* `*self` may dangle */
}
```